utils/src/main/java/software/amazon/awssdk/utils/cache/NonBlocking.java

```diff
@@ -37,80 +46,204 @@
+        }
+
+        Instant maxEnd = rangeStart.plus(1, HOURS);
+        Instant rangeEnd = result.staleTime().isAfter(maxEnd) ? maxEnd : result.staleTime().minus(1, MINUTES);
+
+        if (rangeEnd.isBefore(rangeStart)) {
+            return Optional.of(rangeStart);
+        }
+
+        return Optional.of(randomTimeBetween(rangeStart, rangeEnd));
+    }
+
+    private Instant randomTimeBetween(Instant rangeStart, Instant rangeEnd) {
+        Duration timeBetween = Duration.between(rangeStart, rangeEnd);
+        return rangeStart.plusMillis(Math.abs(JITTER_RANDOM.nextLong() % timeBetween.toMillis()));
+    }
+
+    private void schedulePrefetch(Instant refreshTime) {
+        if (shutdown) {
+            return;
+        }
+
+        Duration waitTime = Duration.between(Instant.now(), refreshTime);
+        log.debug(() -> "Scheduling refresh attempt for " + refreshTime + " (in " + waitTime.toMillis() + " ms)");
+        updateTask(SCHEDULER.schedule(() -> {
+            Thread.currentThread().setName(asyncThreadName + "-scheduler");
+            log.debug(() -> "Executing refresh attempt scheduled for " + refreshTime);
+
+            // If the supplier has already been prefetched, this will just be a cache hit.
+            tryRunBackgroundTask(cachedSupplier::get);
+        }, waitTime.toMillis(), TimeUnit.MILLISECONDS));
+
+        if (shutdown) {
+            updateTask(null);
         }
     }
 
     @Override
     public void close() {
-        executor.shutdown();
+        shutdown = true;
+        updateTask(null);
+    }
+
+    public void updateTask(ScheduledFuture<?> newTask) {
+        ScheduledFuture<?> currentTask;
+        do {
+            currentTask = refreshTask.get();
+            if (currentTask != null && !currentTask.isDone()) {
+                currentTask.cancel(false);
+            }
+        } while (!refreshTask.compareAndSet(currentTask, newTask));
+    }
+
+    public void tryRunBackgroundTask(Runnable runnable) {
+        tryRunBackgroundTask(runnable, () -> {
+        });
+    }
+
+    public void tryRunBackgroundTask(Runnable runnable, Runnable finallyRunnable) {
+        try {
+            if (EXECUTOR.getActiveCount() > MAX_CONCURRENT_REFRESHES) {
+                log.warn(() -> "Skipping a background refresh task because there are too many other tasks running.");
+                return;
+            }
+
+            EXECUTOR.submit(() -> {
+                try {
+                    Thread.currentThread().setName(asyncThreadName);
+                    runnable.run();
+                } catch (Throwable t) {
+                    log.warn(() -> "Exception occurred in AWS SDK background task.", t);
+                } finally {
+                    finallyRunnable.run();
+                }
+            });
+        } catch (Throwable t) {
+            log.warn(() -> "Exception occurred when submitting AWS SDK background task.", t);
+        } finally {
+            finallyRunnable.run();
```

Comment:
finallyRunnable.run() will be called twice - once when registering the task, a second one when executing, forcing that to be an idempotent operation (or introducing subtle bugs).

Perhaps the outer finally can be merged into the catch - either executing it on exception when registering, or whenever executed.

Reply:
This may already be a bug for this caller:

``` 
       if (currentlyPrefetching.compareAndSet(false, true)) {
            tryRunBackgroundTask(valueUpdater, () -> currentlyPrefetching.set(false));
        }
```

Reply:
Yep, it is. I added that bug when I factored out the method. Good catch!

Reply:
Yep, that's what I meant to do originally. Good eye.

Reply:
Fixed.